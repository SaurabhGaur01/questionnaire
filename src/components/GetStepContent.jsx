import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';
import UserName from './CustomerInputs/UserName';
import UserChildren from './CustomerInputs/UserChildren';
import UserEmail from './CustomerInputs/UserEmail';
import UserAddress from './CustomerInputs/UserAddress';
import UserOccupation from './CustomerInputs/UserOccupation';

const GetStepContent = ({ activeStep }) => {
    switch (activeStep) {
        case 0:
            return (
                <UserName />
            );
        case 1:
            return (
                <UserAddress />
            );
        case 2:
            return (
                <UserOccupation />
            );
        case 3:
            return (
                <UserChildren />
            );
        case 4:
            return (
                <UserEmail />
            );
        default:
            return <UserName />;
    }
}

GetStepContent.propTypes = {
    activeStep:  PropTypes.number.isRequired,
};

const mapStateToProps = state => ({
    activeStep: state.activeStep,
});

const hocChain = compose(
    connect(mapStateToProps),
);

export { GetStepContent as TestableGetStepContent };

export default hocChain(GetStepContent);