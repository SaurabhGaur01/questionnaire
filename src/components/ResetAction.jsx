import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';
import handleReset from '../thunks/handleReset';

const ResetAction = ({
    actionHandleReset,
}) => {
    const handleReset = () => {
        actionHandleReset();
    };

    return (
        <div className="action-buttons">
            <button 
                id ="reset-button" 
                type="button" 
                className="btn btn-primary"
                onClick={handleReset}
            >
                Reset
            </button>
        </div>  
    );
}

ResetAction.propTypes = {
    actionHandleReset: PropTypes.func.isRequired,
};

const mapDispatchAsProps = {
    actionHandleReset: handleReset,
}

const hocChain = compose(
    connect(null, mapDispatchAsProps),
);

export { ResetAction as TestableResetAction };

export default hocChain(ResetAction);