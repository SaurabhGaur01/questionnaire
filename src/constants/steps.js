export const QUESTIONNAIRE_STEPS = [
    'Step-1', 
    'Step-2', 
    'Step-3', 
    'Step-4', 
    'Step-5'
];

export const LOADING_ERROR = "ERROR";
export const LOADING_IN_PROGRESS = "LOADING";
export const LOADING_SUCCESS = "SUCCESS";
export const BASE_URL = "https://challenge-dot-popsure-204813.appspot.com";