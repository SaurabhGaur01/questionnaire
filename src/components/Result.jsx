import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { LOADING_ERROR, LOADING_IN_PROGRESS } from '../constants/steps';
import handleReset from '../thunks/handleReset'; 

const Result = ({ 
    searchResults,
    loadingStatus,
    actionHandleReset,
}) => {
    const handleReset = () => {
        actionHandleReset();
    };

    if(loadingStatus === LOADING_ERROR) {
        return (
            <div className="alert alert-danger" role="alert">
                There is an error, Please try again after sometime
            </div>
        )
    }

    if(loadingStatus === LOADING_IN_PROGRESS){
        return (
            <div className="progress">
                <div 
                    className="progress-bar progress-bar-striped bg-success progress-bar" 
                    role="progressbar" 
                    aria-valuenow="25" 
                    aria-valuemin="0" 
                    aria-valuemax="100" />
            </div> 
        );
    }

    return (
        <React.Fragment>
            <div className="card mt-3">
                <div className="card-body">
                    <div className="result-heading">We got your recommendation</div>
                    <div className="result-sub-heading">Based on your answers, this is what make sense for you and what you should pay.</div>
                    {
                        searchResults.map(({ type, price: { amount, periodicity }}, index) => (
                            <div className="card mt-2 mb-2" key={index}>
                                <div className="card-body">
                                    <div className="inner-content">
                                        <div className="row-1">{type.split('_').join(' ')}</div>
                                        <div className="row-1">Â£{amount} per {periodicity.toLowerCase()}</div>
                                    </div>
                                </div>    
                            </div>    
                        ))
                    }
                </div>    
            </div>
        
            <div className="action-buttons">
                <button 
                    id ="reset-button" 
                    type="button" 
                    className="btn btn-primary"
                    onClick={handleReset}
                >
                    Reset
                </button>
            </div>         
        </React.Fragment> 
    );
}    

Result.propTypes = {
    actionHandleReset: PropTypes.func.isRequired,
    searchResults: PropTypes.arrayOf(PropTypes.shape({
        type: PropTypes.string.isRequired,
        price: PropTypes.shape({
            amount: PropTypes.number.isRequired,
            periodicity: PropTypes.string.isRequired,
        })
    })),
    loadingStatus: PropTypes.string.isRequired,
}

export const mapStateToProps = state => ({
    searchResults: state.searchResults,
    loadingStatus: state.loadingStatus,
});

const mapDispatchAsProps = {
    actionHandleReset: handleReset,
}

const hocChain = compose(
    connect(mapStateToProps, mapDispatchAsProps),
);

export { Result as TestableResult };

export default hocChain(Result);