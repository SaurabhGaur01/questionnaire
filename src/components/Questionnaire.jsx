import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { setStepsLength } from '../ducks/stepsLength';
import { setStepAsCompleted } from '../ducks/completedSteps';
import { setActiveStep } from '../ducks/activeStep';
import { QUESTIONNAIRE_STEPS } from '../constants/steps';
import AppStepper from './Shared/AppStepper';
import Result from './Result';
import ActionBar from './Shared/ActionBar';
import handleResult from '../thunks/handleResult';
import GetStepContent from './GetStepContent';

const Questionnaire = ({ 
    stepsLength,
    activeStep, 
    actionSetStepsLength,
    completedSteps,
    actionSetStepAsCompleted,
    actionSetActiveStep,
    actionHandleResult, 
}) => {
    React.useEffect(() => {
        // getting all steps required for Questionnaire 
        // and set FirstName as a active step for persistance
        actionSetStepsLength();
        actionSetActiveStep(0);
    }, [actionSetStepsLength, actionSetActiveStep]);

    const allStepsCompleted = () => {
        return Object.keys(completedSteps).length === stepsLength;
    };

    const handleNext = () => {
        const newActiveStep = (activeStep === stepsLength - 1) 
        && !(allStepsCompleted())
        ? // It's the last step, but not all steps have been completed,
        // find the first step that has been completed
        QUESTIONNAIRE_STEPS.findIndex((step, i) => !(i in completedSteps))
        : activeStep + 1;
        actionSetActiveStep(newActiveStep);

        // if all steps completed, make an API call
        if(allStepsCompleted()) {
            actionHandleResult();
        }
    };
    
    const handleComplete = () => {
        const newCompleted = completedSteps;
        // updating object with completed ones
        newCompleted[activeStep] = true;
        actionSetStepAsCompleted(newCompleted);
        handleNext();
    };

    return(
        <React.Fragment>
            <AppStepper steps={QUESTIONNAIRE_STEPS} />
            {
                allStepsCompleted() ? (
                    <Result allStepsCompleted={allStepsCompleted} />
                ) : (
                <React.Fragment>
                    <div className="row">
                        <div className="container">
                            <div className="col-md-4"></div>
                            <div className="col-md-6">
                                <GetStepContent />
                                <ActionBar handleComplete={handleComplete} />
                            </div>
                        </div>
                    </div>
                </React.Fragment>
                )
            }
        </React.Fragment>
    );
}    

Questionnaire.propTypes = {
    stepsLength: PropTypes.number.isRequired,
    activeStep: PropTypes.number.isRequired,
    actionSetStepsLength: PropTypes.func.isRequired,
    actionSetStepAsCompleted: PropTypes.func.isRequired,
    actionSetActiveStep: PropTypes.func.isRequired,
    completedSteps: PropTypes.shape({}).isRequired,
    actionHandleResult: PropTypes.func.isRequired,
};

// Actions dispatchers
const mapDispatchAsProps = {
    actionSetStepsLength: setStepsLength,
    actionSetStepAsCompleted: setStepAsCompleted,
    actionSetActiveStep: setActiveStep,
    actionHandleResult: handleResult,
}

// Redux store access
export const mapStateToProps = state => ({
    stepsLength: state.stepsLength,
    completedSteps: state.completedSteps,
    activeStep: state.activeStep,
});

const hocChain = compose(
    connect(mapStateToProps, mapDispatchAsProps),
);

// this is required for creating a wrapper for test coverage
export { Questionnaire as TestableQuestionnaire };

export default hocChain(Questionnaire);