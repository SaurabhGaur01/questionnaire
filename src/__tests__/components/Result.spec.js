import React from 'react';
import { shallow } from 'enzyme';
import { TestableResult, mapStateToProps } from '../../components/Result';
import { LOADING_ERROR, LOADING_IN_PROGRESS, LOADING_SUCCESS } from '../../constants/steps';

describe('Result Component', () => {
    let props;

    beforeEach(() =>{
        props = {
            actionHandleReset: jest.fn(),
            loadingStatus: LOADING_SUCCESS,
            searchResults: [{
                type: 'mockType',
                price: {
                    amount: 100,
                    periodicity: 'month'
                }
            }]
        };
    });

    it('should render correctly', () => {
        const renderedModule = shallow(<TestableResult {...props} />);
        expect(renderedModule).toMatchSnapshot();
    });

    it('should render correctly when loadingStatus is in Progress', () => {
        props.loadingStatus = LOADING_IN_PROGRESS;
        const renderedModule = shallow(<TestableResult {...props} />);
        expect(renderedModule).toMatchSnapshot();
    });

    it('should render correctly when loadingStatus is in Error', () => {
        props.loadingStatus = LOADING_ERROR;
        const renderedModule = shallow(<TestableResult {...props} />);
        expect(renderedModule).toMatchSnapshot();
    });

    it('should call actionHandleReset when RESET button is clicked', () => {
        const renderedModule = shallow(<TestableResult {...props} />);
        renderedModule.find('#reset-button').at(0).simulate('click');
        expect(props.actionHandleReset).toHaveBeenCalledTimes(1);
    });

    describe('mapStateToProps', () => {
        const loadingStatusMock = 'loadingStatusMock';
        const searchResultsMock = [{
            mockkey: 'mockValue',
        }]

        const mockedState = {
            searchResults: searchResultsMock,
            loadingStatus: loadingStatusMock,
        };

        it('should return the module state from the redux store', () => {
            const mockPropsFromRedux = mapStateToProps(mockedState);
            expect(mockPropsFromRedux).toEqual(
                {
                    searchResults: searchResultsMock,
                    loadingStatus: loadingStatusMock
                }
            )
        })
    })
})