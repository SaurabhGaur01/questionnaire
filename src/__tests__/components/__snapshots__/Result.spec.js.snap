// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result Component should render correctly 1`] = `
<Fragment>
  <div
    className="card mt-3"
  >
    <div
      className="card-body"
    >
      <div
        className="result-heading"
      >
        We got your recommendation
      </div>
      <div
        className="result-sub-heading"
      >
        Based on your answers, this is what make sense for you and what you should pay.
      </div>
      <div
        className="card mt-2 mb-2"
        key="0"
      >
        <div
          className="card-body"
        >
          <div
            className="inner-content"
          >
            <div
              className="row-1"
            >
              mockType
            </div>
            <div
              className="row-1"
            >
              Â£
              100
               per 
              month
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="action-buttons"
  >
    <button
      className="btn btn-primary"
      id="reset-button"
      onClick={[Function]}
      type="button"
    >
      Reset
    </button>
  </div>
</Fragment>
`;

exports[`Result Component should render correctly when loadingStatus is in Error 1`] = `
<div
  className="alert alert-danger"
  role="alert"
>
  There is an error, Please try again after sometime
</div>
`;

exports[`Result Component should render correctly when loadingStatus is in Progress 1`] = `
<div
  className="progress"
>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="25"
    className="progress-bar progress-bar-striped bg-success progress-bar"
    role="progressbar"
  />
</div>
`;
